### DISCUSSION TITLE
How to export bottom drag?

### DISCUSSION THREAD

**Discussion by jhdong2016** [CliMA/Oceananigans.jl#3081] [Computational science!] [answered]
I am running a simulation focusing on the bottom mixed layer. How can I export bottom drag if I am interested about that?

**Comment by glwagner** [comment, marked as answer]
Unfortunately there is no user interface to build `Field` from boundary conditions, though we do need one. It's still possible, but a bit of work. This code is fairly general I think:

```julia
using Oceananigans.BoundaryConditions: getbc
using Oceananigans: fields

# Boundary condition extractor in "kernel function form"
@inline kernel_getbc(i, j, k, grid, boundary_condition, clock, fields) =
    getbc(boundary_condition, i, j, grid, clock, fields)

# Kernel arguments
grid = model.grid
clock = model.clock
model_fields = merge(fields(model), model.auxiliary_fields)
u, v, w = model.velocities
u_bc = u.boundary_conditions.bottom
v_bc = v.boundary_conditions.bottom

# Build operations
u_bc_op = KernelFunctionOperation{Face, Center, Nothing}(kernel_getbc, grid, u_bc, clock, model_fields)
v_bc_op = KernelFunctionOperation{Center, Face, Nothing}(kernel_getbc, grid, v_bc, clock, model_fields)
```

`u_bc_op` and `v_bc_op` can be passed to `OutputWriters`. You can also build `Field`s from them:

```julia
# Build Fields
u_bc_field = Field(u_bc_op)
v_bc_field = Field(v_bc_op)
```

Note that this will compute the boundary condition directly. So if you specify a boundary condition as `FluxBoundaryCondition`, this outputs the flux. If you are using `ValueBoundaryCondition`, however, you would not get the momentum flux from this code. For that we need to work a little harder...

Another possibility is simply to write a function that computes the drag as it's been specified. However to do that correctly you need to be careful to handle the staggered grid correctly. If you share your code, we can help with that.

**Reply by jhdong2016** [reply]
Thanks for your reply! I will try that for my simulation!

**Reply by jhdong2016** [reply]
I use the code as you suggested, but it reported an error as follows. Do you have any idea for solving it?

![image](https://user-images.githubusercontent.com/46232007/233816760-ba76310a-6ab8-469a-aeaf-f447cddee97e.png)

**Reply by simone-silvestri** [reply]
Hello. That error shows that you are using an order version of Oceananigans where the signature  for  `KernelFunctionOperation` was different. You can use the old version and write:
```
u_bc_op=KernelFunctionOperation{Face, Center, Nothing}(kernel_getbc, grid; computed_dependencies=(u_bc, clock, model_fields))
v_bc_op=KernelFunctionOperation{Center, Face, Nothing}(kernel_getbc, grid; computed_dependencies=(v_bc, clock, model_fields))
```
Or you can update Oceananigans to the newer version 
```
using Pkg
Pkg.update("Oceananigans")
```

**Comment by josuemtzmo** [comment]
Hello, I'm also looking into this, is there a new user interface to build the `Field` for the boundary conditions without the need of this cool brute force way?

**Reply by glwagner** [reply]
There's nothing automatic, but you could open an issue that outlines the feature you think would be helpful, then we can come up with what the syntax should look like, and then we can implement it.
